components/log/log.c

```diff
@@ -0,0 +1,159 @@
+#include "log.h"
+
+#include <errno.h>
+#include <string.h>
+#include <unistd.h>
+
+// The names are long, so let's add a convenience helpers
+typedef datadog_php_logger logger_t;
+typedef datadog_php_log_level log_level_t;
+typedef datadog_php_string_view string_view_t;
+
+logger_t datadog_php_logger_ctor(int descriptor, log_level_t log_level, pthread_mutex_t *mutex) {
+    logger_t logger = {
+        .descriptor = descriptor < 0 ? -1 : descriptor,  // normalize invalid
+        .log_level = log_level,
+        .mutex = mutex,
+    };
+    return logger;
+}
+
+static bool logger_valid(datadog_php_logger *logger) {
+    return logger->descriptor >= 0 && logger->log_level >= DATADOG_PHP_LOG_OFF &&
+           logger->log_level <= DATADOG_PHP_LOG_DEBUG && logger->mutex;
+}
+
+bool datadog_php_logger_valid(datadog_php_logger *logger) {
+    bool valid = false;
+    if (logger->mutex && pthread_mutex_lock(logger->mutex) == 0) {
+        valid = logger_valid(logger);
+        (void)pthread_mutex_unlock(logger->mutex);
+    }
+    return valid;
+}
```

Comment:
This API does not make much sense. Either you want - as a caller - lock yourself and then do some log operation. Or you generally don't care about the thread safety of this.
But here, if you were to mutex this, as a caller, and then use this function, you'll deadlock yourself.

Reply:
I didn't understand your feedback, but I do agree it's a poor function and can be removed because the profiler only uses is in these ways:

 1. Detect that initialization failed. We can make initialization return this info somehow.
 2. Avoid double-dtors. We can handle this differently.
 
 Sound good?

Reply:
Sounds good to me.